File: examples/crypto/kzg/data-availability.ts
/**
 * Data Availability Sampling Example
 *
 * Demonstrates:
 * - L2 rollup data availability workflow
 * - Random point sampling and verification
 * - Light client data availability checks
 * - Understanding DA guarantees
 */

import { KZG } from '../../../src/crypto/KZG/KZG.js';
import { SHA256 } from '../../../src/crypto/SHA256/index.js';
import * as Hex from '../../../src/primitives/Hex/index.js';

console.log('=== Data Availability Sampling ===\n');

// Initialize KZG
KZG.loadTrustedSetup();

// Step 1: L2 sequencer posts data
console.log('1. L2 Sequencer: Publishing Rollup Data');
console.log('-'.repeat(50));

// Simulate L2 batch
const rollupBatch = {
  batchNumber: 12345,
  transactions: 1500,
  compressedSize: 128 * 1024, // 128 KB
  stateRoot: '0x1234...5678',
};

console.log('L2 Batch:');
console.log('  Batch number:', rollupBatch.batchNumber);
console.log('  Transactions:', rollupBatch.transactions);
console.log('  Compressed size:', rollupBatch.compressedSize, 'bytes');
console.log('  State root:', rollupBatch.stateRoot);
console.log();

// Create blob from rollup data
const blob = KZG.generateRandomBlob(); // In reality: encode compressed batch
const commitment = KZG.blobToKzgCommitment(blob);
const versionedHash = (() => {
  const hash = SHA256.hash(commitment);
  hash[0] = 0x01;
  return hash;
})();

console.log('Publishing to L1:');
console.log('  Blob size:', blob.length, 'bytes');
console.log('  Commitment:', Hex.fromBytes(commitment).slice(0, 20) + '...');
console.log('  Versioned hash:', Hex.fromBytes(versionedHash));
console.log('  Status: ✓ Submitted to L1');
console.log();

// Step 2: Full nodes download and verify
console.log('2. Full Nodes: Download and Verify');
console.log('-'.repeat(50));

console.log('Full node workflow:');
console.log('  1. Download blob from consensus layer');
console.log('  2. Compute KZG commitment from blob');
console.log('  3. Verify commitment matches versioned hash');
console.log('  4. Store blob for 18-day availability window');
console.log();

// Verify commitment matches
const recomputedCommitment = KZG.blobToKzgCommitment(blob);
const commitmentMatches = recomputedCommitment.every((b, i) => b === commitment[i]);

console.log('Full node verification:');
console.log('  Downloaded blob: ✓');
console.log('  Recomputed commitment: ✓');
console.log('  Commitment matches:', commitmentMatches ? '✓ Yes' : '✗ No');
console.log('  Data available: ✓');
console.log();

// Step 3: Light client sampling
console.log('3. Light Client: Data Availability Sampling');
console.log('-'.repeat(50));

console.log('Light client strategy:');
console.log('  - Cannot download full 128 KB blob');
console.log('  - Samples random points via KZG proofs');
console.log('  - Verifies each sample against commitment');
console.log('  - Statistical confidence in data availability');
console.log();

// Generate random sample points
const NUM_SAMPLES = 5;
const samplePoints: Uint8Array[] = [];

console.log(`Generating ${NUM_SAMPLES} random sample points...`);
for (let i = 0; i < NUM_SAMPLES; i++) {
  const z = new Uint8Array(32);
  crypto.getRandomValues(z);
  z[0] = 0; // Ensure < BLS modulus
  samplePoints.push(z);
}
console.log();

// Step 4: Request proofs from full node
console.log('4. Requesting Proofs from Full Node');
console.log('-'.repeat(50));

const proofs: Array<{ z: Uint8Array; y: Uint8Array; proof: Uint8Array }> = [];

console.log('Light client requests proofs for sample points:');
for (let i = 0; i < NUM_SAMPLES; i++) {
  const z = samplePoints[i];
  const { proof, y } = KZG.computeKzgProof(blob, z);

  proofs.push({ z, y, proof });

  console.log(`  Sample ${i + 1}:`);
  console.log('    z:', Hex.fromBytes(z).slice(0, 16) + '...');
  console.log('    y:', Hex.fromBytes(y).slice(0, 16) + '...');
  console.log('    proof size:', proof.length, 'bytes');
}
console.log();

console.log('Data transferred to light client:');
console.log('  Commitment:', 48, 'bytes');
console.log('  Proofs:', NUM_SAMPLES * (32 + 48), 'bytes (z + proof)');
console.log('  Y values:', NUM_SAMPLES * 32, 'bytes');
console.log('  Total:', 48 + NUM_SAMPLES * (32 + 48 + 32), 'bytes');
console.log('  vs Full blob:', blob.length, 'bytes');
console.log('  Savings:', ((1 - (48 + NUM_SAMPLES * 112) / blob.length) * 100).toFixed(1) + '%');
console.log();

// Step 5: Light client verifies samples
console.log('5. Light Client: Verifying Samples');
console.log('-'.repeat(50));

let allValid = true;
console.log('Verifying each sample proof:');

for (let i = 0; i < proofs.length; i++) {
  const { z, y, proof } = proofs[i];
  const valid = KZG.verifyKzgProof(commitment, z, y, proof);

  console.log(`  Sample ${i + 1}: ${valid ? '✓' : '✗'}`);

  if (!valid) {
    allValid = false;
  }
}
console.log();

console.log('All samples valid:', allValid ? '✓ Yes' : '✗ No');
console.log('Data availability confidence:', allValid ? 'High' : 'Low');
console.log();

// Step 6: Understanding DA guarantees
console.log('6. Data Availability Guarantees');
console.log('-'.repeat(50));

console.log('What KZG proves:');
console.log('  ✓ Data exists (blob was published)');
console.log('  ✓ Commitment binds to specific data');
console.log('  ✓ Cannot change data after commitment');
console.log('  ✓ Samples prove polynomial evaluation');
console.log();

console.log('What KZG does NOT prove:');
console.log('  ✗ Data correctness (state transition validity)');
console.log('  ✗ Transaction ordering');
console.log('  ✗ Smart contract execution');
console.log('  → These require fraud/validity proofs');
console.log();

// Step 7: Attack scenarios
console.log('7. Attack Scenarios');
console.log('-'.repeat(50));

console.log('Scenario 1: Malicious sequencer withholds blob');
console.log('  - Commitment on-chain, but blob not published');
console.log('  - Light client cannot get valid proofs');
console.log('  - Full nodes detect missing blob');
console.log('  - L2 halts or falls back to L1 data');
console.log('  Result: ✓ Attack detected');
console.log();

console.log('Scenario 2: Sequencer provides fake proof');
console.log('  - Tries to fool light client with wrong y value');
const fakeY = new Uint8Array(32);
crypto.getRandomValues(fakeY);
fakeY[0] = 0;
const fakeValid = KZG.verifyKzgProof(commitment, samplePoints[0], fakeY, proofs[0].proof);
console.log('  - Fake proof verification:', fakeValid ? '✗ Passed (BAD)' : '✓ Failed (expected)');
console.log('  Result: ✓ Attack prevented by crypto');
console.log();

console.log('Scenario 3: Sequencer publishes invalid state transition');
console.log('  - Blob is available, but state transition wrong');
console.log('  - KZG proves DA, not correctness');
console.log('  - Requires fraud proof (optimistic) or validity proof (ZK)');
console.log('  Result: ✓ Caught by fraud/validity proof system');
console.log();

// Step 8: Sampling statistics
console.log('8. Sampling Statistics');
console.log('-'.repeat(50));

function calculateConfidence(numSamples: number): number {
  // Simplified: probability of detecting unavailable data
  // In reality, depends on erasure coding and Reed-Solomon properties
  return 1 - Math.pow(0.5, numSamples);
}

const confidenceLevels = [1, 5, 10, 20, 50];

console.log('Data availability confidence by sample count:');
for (const n of confidenceLevels) {
  const confidence = calculateConfidence(n);
  console.log(`  ${n.toString().padStart(2)} samples: ${(confidence * 100).toFixed(2)}%`);
}
console.log();

console.log('Trade-off:');
console.log('  More samples → Higher confidence');
console.log('  More samples → More bandwidth for light client');
console.log('  Typical: 5-10 samples for light clients');
console.log();

// Step 9: Full DA workflow
console.log('9. Complete DA Workflow');
console.log('-'.repeat(50));

console.log('Timeline:');
console.log();

console.log('Block N (blob published):');
console.log('  L2 → Publishes blob + commitment to L1');
console.log('  L1 → Includes blob transaction in block');
console.log('  Consensus → Stores blob in sidecar');
console.log('  Full nodes → Download and verify blob');
console.log();

console.log('Blocks N+1 to N+4096 (~18 days):');
console.log('  Full nodes → Store blob');
console.log('  Light clients → Sample on demand');
console.log('  Archive nodes → May keep longer');
console.log('  Challengers → Can submit fraud proofs');
console.log();

console.log('After block N+4096:');
console.log('  Consensus → Prunes blob (not commitment!)');
console.log('  Commitment → Remains in state forever');
console.log('  Historical access → Only via archive nodes');
console.log();

// Step 10: Production recommendations
console.log('10. Production Recommendations');
console.log('-'.repeat(50));

console.log('For L2 operators:');
console.log('  ✓ Publish blobs with every batch');
console.log('  ✓ Store blobs locally (don\'t rely on 18-day window)');
console.log('  ✓ Provide DA proof endpoints for light clients');
console.log('  ✓ Monitor blob inclusion and availability');
console.log();

console.log('For light clients:');
console.log('  ✓ Sample 5-10 random points per blob');
console.log('  ✓ Request proofs from multiple full nodes');
console.log('  ✓ Verify all samples before trusting DA');
console.log('  ✓ Fall back to L1 data if samples fail');
console.log();

console.log('For dapp developers:');
console.log('  ✓ Understand DA != correctness');
console.log('  ✓ Rely on fraud/validity proofs for correctness');
console.log('  ✓ Monitor L2 liveness and DA');
console.log('  ✓ Have escape hatch for DA failures');
console.log();

// Cleanup
KZG.freeTrustedSetup();

console.log('=== Key Takeaways ===');
console.log('- KZG enables efficient data availability sampling');
console.log('- Light clients verify DA without downloading full blob');
console.log('- Sample 5-10 random points for statistical confidence');
console.log('- DA proves existence, not correctness');
console.log('- Critical for L2 scaling and decentralization');
console.log('- Enables Proto-Danksharding vision');

File: packages/voltaire-zig/examples/crypto/kzg/data-availability.zig
// Data Availability Sampling Example
//
// Demonstrates:
// - L2 rollup data availability workflow
// - Random point sampling and verification
// - Light client data availability checks
// - Understanding DA guarantees

const std = @import("std");
const crypto = @import("crypto");

const c_kzg = crypto.c_kzg;
const Sha256 = crypto.sha256.Sha256;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n=== Data Availability Sampling ===\n\n", .{});

    // Initialize KZG
    const ckzg = @import("c_kzg");
    try ckzg.loadTrustedSetupFromText(ckzg.embedded_trusted_setup, 0);
    defer ckzg.freeTrustedSetup() catch {};

    // Step 1: L2 sequencer posts data
    std.debug.print("1. L2 Sequencer: Publishing Rollup Data\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("L2 Batch:\n", .{});
    std.debug.print("  Batch number: 12345\n", .{});
    std.debug.print("  Transactions: 1500\n", .{});
    std.debug.print("  Compressed size: 131072 bytes\n", .{});
    std.debug.print("  State root: 0x1234...5678\n\n", .{});

    var blob = getRandomBlob(12345);
    const commitment = try c_kzg.blobToKZGCommitment(&blob);

    var versioned_hash: [32]u8 = undefined;
    Sha256.hash(&commitment, &versioned_hash);
    versioned_hash[0] = 0x01;

    std.debug.print("Publishing to L1:\n", .{});
    std.debug.print("  Blob size: {} bytes\n", .{blob.len});
    std.debug.print("  Commitment: 0x", .{});
    for (commitment[0..10]) |b| {
        std.debug.print("{x:0>2}", .{b});
    }
    std.debug.print("...\n", .{});
    std.debug.print("  Versioned hash: 0x", .{});
    for (versioned_hash) |b| {
        std.debug.print("{x:0>2}", .{b});
    }
    std.debug.print("\n", .{});
    std.debug.print("  Status: ✓ Submitted to L1\n\n", .{});

    // Step 2: Full nodes verify
    std.debug.print("2. Full Nodes: Download and Verify\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Full node workflow:\n", .{});
    std.debug.print("  1. Download blob from consensus layer\n", .{});
    std.debug.print("  2. Compute KZG commitment from blob\n", .{});
    std.debug.print("  3. Verify commitment matches versioned hash\n", .{});
    std.debug.print("  4. Store blob for 18-day availability window\n\n", .{});

    const recomputed_commitment = try c_kzg.blobToKZGCommitment(&blob);
    const commitment_matches = std.mem.eql(u8, &recomputed_commitment, &commitment);

    std.debug.print("Full node verification:\n", .{});
    std.debug.print("  Downloaded blob: ✓\n", .{});
    std.debug.print("  Recomputed commitment: ✓\n", .{});
    std.debug.print("  Commitment matches: {s}\n", .{if (commitment_matches) "✓ Yes" else "✗ No"});
    std.debug.print("  Data available: ✓\n\n", .{});

    // Step 3: Light client sampling
    std.debug.print("3. Light Client: Data Availability Sampling\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Light client strategy:\n", .{});
    std.debug.print("  - Cannot download full 128 KB blob\n", .{});
    std.debug.print("  - Samples random points via KZG proofs\n", .{});
    std.debug.print("  - Verifies each sample against commitment\n", .{});
    std.debug.print("  - Statistical confidence in data availability\n\n", .{});

    const NUM_SAMPLES = 5;
    var sample_points: [NUM_SAMPLES]c_kzg.Bytes32 = undefined;

    std.debug.print("Generating {} random sample points...\n\n", .{NUM_SAMPLES});
    var prng = std.Random.DefaultPrng.init(99999);
    const random = prng.random();

    for (0..NUM_SAMPLES) |i| {
        random.bytes(&sample_points[i]);
        sample_points[i][0] = 0;
    }

    // Step 4: Request proofs
    std.debug.print("4. Requesting Proofs from Full Node\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Light client requests proofs for sample points:\n", .{});

    var proofs: [NUM_SAMPLES]struct { z: c_kzg.Bytes32, y: c_kzg.Bytes32, proof: c_kzg.KZGProof } = undefined;

    for (0..NUM_SAMPLES) |i| {
        const z = sample_points[i];
        const result = try c_kzg.computeKZGProof(&blob, &z);

        proofs[i] = .{ .z = z, .y = result.y, .proof = result.proof };

        std.debug.print("  Sample {}:\n", .{i + 1});
        std.debug.print("    z: 0x", .{});
        for (z[0..8]) |b| {
            std.debug.print("{x:0>2}", .{b});
        }
        std.debug.print("...\n", .{});
        std.debug.print("    y: 0x", .{});
        for (result.y[0..8]) |b| {
            std.debug.print("{x:0>2}", .{b});
        }
        std.debug.print("...\n", .{});
        std.debug.print("    proof size: {} bytes\n", .{result.proof.len});
    }
    std.debug.print("\n", .{});

    const data_transferred = 48 + NUM_SAMPLES * (32 + 48 + 32);
    std.debug.print("Data transferred to light client:\n", .{});
    std.debug.print("  Commitment: 48 bytes\n", .{});
    std.debug.print("  Proofs: {} bytes (z + proof)\n", .{NUM_SAMPLES * (32 + 48)});
    std.debug.print("  Y values: {} bytes\n", .{NUM_SAMPLES * 32});
    std.debug.print("  Total: {} bytes\n", .{data_transferred});
    std.debug.print("  vs Full blob: {} bytes\n", .{blob.len});

    const savings_percent = (1.0 - (@as(f64, @floatFromInt(data_transferred)) / @as(f64, @floatFromInt(blob.len)))) * 100.0;
    std.debug.print("  Savings: {d:.1}%\n\n", .{savings_percent});

    // Step 5: Verify samples
    std.debug.print("5. Light Client: Verifying Samples\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    var all_valid = true;
    std.debug.print("Verifying each sample proof:\n", .{});

    for (0..NUM_SAMPLES) |i| {
        const valid = try c_kzg.verifyKZGProof(&commitment, &proofs[i].z, &proofs[i].y, &proofs[i].proof);
        std.debug.print("  Sample {}: {s}\n", .{ i + 1, if (valid) "✓" else "✗" });

        if (!valid) {
            all_valid = false;
        }
    }
    std.debug.print("\n", .{});

    std.debug.print("All samples valid: {s}\n", .{if (all_valid) "✓ Yes" else "✗ No"});
    std.debug.print("Data availability confidence: {s}\n\n", .{if (all_valid) "High" else "Low"});

    // Step 6: DA guarantees
    std.debug.print("6. Data Availability Guarantees\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("What KZG proves:\n", .{});
    std.debug.print("  ✓ Data exists (blob was published)\n", .{});
    std.debug.print("  ✓ Commitment binds to specific data\n", .{});
    std.debug.print("  ✓ Cannot change data after commitment\n", .{});
    std.debug.print("  ✓ Samples prove polynomial evaluation\n\n", .{});

    std.debug.print("What KZG does NOT prove:\n", .{});
    std.debug.print("  ✗ Data correctness (state transition validity)\n", .{});
    std.debug.print("  ✗ Transaction ordering\n", .{});
    std.debug.print("  ✗ Smart contract execution\n", .{});
    std.debug.print("  → These require fraud/validity proofs\n\n", .{});

    // Step 7: Sampling statistics
    std.debug.print("7. Sampling Statistics\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    const confidence_levels = [_]u32{ 1, 5, 10, 20, 50 };

    std.debug.print("Data availability confidence by sample count:\n", .{});
    for (confidence_levels) |n| {
        const confidence = 1.0 - std.math.pow(f64, 0.5, @as(f64, @floatFromInt(n)));
        std.debug.print("  {} samples: {d:.2}%\n", .{ n, confidence * 100.0 });
    }
    std.debug.print("\n", .{});

    std.debug.print("Trade-off:\n", .{});
    std.debug.print("  More samples → Higher confidence\n", .{});
    std.debug.print("  More samples → More bandwidth for light client\n", .{});
    std.debug.print("  Typical: 5-10 samples for light clients\n\n", .{});

    // Step 8: Production recommendations
    std.debug.print("8. Production Recommendations\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("For L2 operators:\n", .{});
    std.debug.print("  ✓ Publish blobs with every batch\n", .{});
    std.debug.print("  ✓ Store blobs locally (don't rely on 18-day window)\n", .{});
    std.debug.print("  ✓ Provide DA proof endpoints for light clients\n", .{});
    std.debug.print("  ✓ Monitor blob inclusion and availability\n\n", .{});

    std.debug.print("For light clients:\n", .{});
    std.debug.print("  ✓ Sample 5-10 random points per blob\n", .{});
    std.debug.print("  ✓ Request proofs from multiple full nodes\n", .{});
    std.debug.print("  ✓ Verify all samples before trusting DA\n", .{});
    std.debug.print("  ✓ Fall back to L1 data if samples fail\n\n", .{});

    std.debug.print("=== Key Takeaways ===\n", .{});
    std.debug.print("- KZG enables efficient data availability sampling\n", .{});
    std.debug.print("- Light clients verify DA without downloading full blob\n", .{});
    std.debug.print("- Sample 5-10 random points for statistical confidence\n", .{});
    std.debug.print("- DA proves existence, not correctness\n", .{});
    std.debug.print("- Critical for L2 scaling and decentralization\n", .{});
    std.debug.print("- Enables Proto-Danksharding vision\n", .{});

    _ = allocator;
}

fn getRandomBlob(seed: u64) c_kzg.Blob {
    var blob: c_kzg.Blob = undefined;
    var prng = std.Random.DefaultPrng.init(seed);
    const random = prng.random();
    random.bytes(&blob);
    for (0..c_kzg.FIELD_ELEMENTS_PER_BLOB) |i| {
        blob[i * c_kzg.BYTES_PER_FIELD_ELEMENT] = 0;
    }
    return blob;
}
